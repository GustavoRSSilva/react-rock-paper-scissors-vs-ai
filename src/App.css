.App {
  text-align: center;
  min-height: 1000px;
}

.App p {
  font-size: large;
}

.App-header {
  background-color: blue;
  height: 100px;
  padding: 20px;
  color: white;
}

.App-title {
  font-size: 1.5em;
}

.App-result-title {
  font-size: 3em;
}

.large{
  font-size: 40px;
}


#player-one-container button {
  margin: 0 10px;
}

#player-one-container button:hover {
  cursor: pointer;
}

.App-error-log{
  color:red;
  margin: 20px 0;
  font-size: 24px
}

.slide-container{
  margin: 0 auto;
  width: 40px;
}

.restart-btn{
  margin: 20px;
  height: 40px;
}

.restart-btn:hover {
  cursor: pointer;
}

.app-footer {
  margin-top: 40px;
}
