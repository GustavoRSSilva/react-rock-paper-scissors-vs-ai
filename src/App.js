import React, { Component } from 'react';
import 'font-awesome/css/font-awesome.min.css';
import './App.css';
import Carousel from './components/Carousel';

const moves = {
  rock: {
    beats: 'scissors',
  },
  scissors: {
    beats: 'paper',
  },
  paper: {
    beats: 'rock',
  }
}

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      playerOneMove: null,
      playerTwoMove: null,
      errorLog: null,
      result: null,
    };
  }

  resetGame() {
    this.setState({
      playerOneMove: null,
      playerTwoMove: null,
      errorLog: null,
      result: null,
    });
  }

  validMove(target) {
    return target && moves[target];
  }

  setMoves(playerOneMove, playerTwoMove = null) {
    //If the Player one move is not valid, we show a error message
    if(!this.validMove(playerOneMove)) {
      return this.setState({ errorLog: 'Invalid move' });
    }

    //If the Player two move is not valid, we show a error message
    if(playerTwoMove !== null && !this.validMove(playerTwoMove)) {
      return this.setState({ errorLog: 'Invalid move' });
    }

    //In order to not override a function parameter, we create an auxiliary variable
    let _playerTwoMove = playerTwoMove;
    const availableMoves = Object.keys(moves);

    //If player two move is null, we set it up
    if(_playerTwoMove === null) {
      _playerTwoMove = availableMoves[Math.floor(Math.random() * availableMoves.length)];
    }

    this.setState(
      { playerOneMove, playerTwoMove: _playerTwoMove },
      () => this.setGameResult()
    );
  }

  setGameResult() {
    const { playerOneMove, playerTwoMove } = this.state;

    if(!this.validMove(playerOneMove) || !this.validMove(playerTwoMove)) {
      return this.setState({ errorLog: 'Invalid operation!' });
    }

    switch (playerOneMove) {
      case playerTwoMove:
        return this.setState({ result: 'It\'s a draw' });

      case moves[playerTwoMove].beats:
        return this.setState({ result: 'You lose!' });

      default:
        return this.setState({ result: 'You win!' });
    }
  }

  renderResult() {
    return (
      <div className="App">
        <header>
          <h1 className="App-result-title">{this.state.result}</h1>
          <h2>
            <p>You picked <i className={`fa fa-hand-${this.state.playerOneMove}-o large`} /></p>
            <p>Vs</p>
            <p>Your opponent picked <i className={`fa fa-hand-${this.state.playerTwoMove}-o large`} /></p>
          </h2>
        </header>
        <button
          className="restart-btn"
          onClick={() => this.resetGame()}
        >
          Restart Game
        </button>
      </div>
    );
  }

  render() {
    // If the game over show the result and the restart button
    if(this.state.result) {
      return this.renderResult();
    }

    return (
      <div className="App">
        <header className="App-header">
          <h1 className="App-title">Rock, paper, scissors</h1>
          <h2>vs AI</h2>
        </header>
        <p>
          Please choose a move.
        </p>
        <div id="player-one-container">
          <button
            className="rock-btn yoo"
            onClick={() => this.setMoves('rock')}
          >
            <i className="fa fa-hand-rock-o large"></i>
          </button>
          <button
            className="scissors-btn"
            onClick={() => this.setMoves('scissors')}
          >
            <i className="fa fa-hand-scissors-o large"></i>
          </button>
          <button
            className="paper-btn"
            onClick={() => this.setMoves('paper')}
          >
            <i className="fa fa-hand-paper-o large"></i>
          </button>
        </div>
        <p>
          vs
        </p>
        <Carousel
          options={Object.keys(moves).map((move) => (<i className={`fa fa-hand-${move}-o large`} />))}
        />
        <p className="App-error-log">
          {this.state.errorLog}
        </p>
        <p className="app-footer">
          For more information about the project, see the <a href="https://gustavorsilva.github.io/blog/react-rock-paper-scissors/" target="_blank">blog post</a>.
        </p>
      </div>
    );
  }
}

export default App;
