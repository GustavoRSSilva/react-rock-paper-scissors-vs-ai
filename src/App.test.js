import React from 'react';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import toJson from 'enzyme-to-json';
import App from './App';


describe('<App />', () => {
  it('renders 1 <App /> component', () => {
    const component = shallow(<App />);
    expect(component).toHaveLength(1);
  })

  it('It renders the state correctly', () => {
    const component = mount(<App />);
    expect(component.state()).toEqual({
      playerOneMove: null,
      playerTwoMove: null,
      errorLog: null,
      result: null,
    });
  })

  it('The Player picks rock, set the player move to rock', () => {
    const component = mount(<App />);
    const rockButton = component.find('button.rock-btn');
    expect(toJson(rockButton)).toMatchSnapshot();

    rockButton.simulate('click');
    expect(component.state().playerOneMove).toBe('rock');
  })

  it('The Player picks paper, set the player move to paper', () => {
    const component = mount(<App />);
    const paperButton = component.find('button.paper-btn');
    expect(toJson(paperButton)).toMatchSnapshot();

    paperButton.simulate('click');
    expect(component.state().playerOneMove).toBe('paper');
  })

  it('The Player picks scissors, set the player move to scissors', () => {
    const component = mount(<App />);
    const scissorsButton = component.find('button.scissors-btn');
    expect(toJson(scissorsButton)).toMatchSnapshot();

    scissorsButton.simulate('click');
    expect(component.state().playerOneMove).toBe('scissors');
  })

  it('The Ai picks a valid move after the player', () => {
    const component = mount(<App />);
    const availableMoves = ['rock', 'paper', 'scissors'];

    const scissorsButton = component.find('button.rock-btn');
    scissorsButton.simulate('click');

    expect(availableMoves).toContain(component.state().playerTwoMove);
  })

  it('The game ends and show the result after both players set the moves', () => {
    const component = mount(<App />);
    const scissorsButton = component.find('button.rock-btn');
    scissorsButton.simulate('click');

    expect(component.state().result).toBeTruthy();
  })
})
