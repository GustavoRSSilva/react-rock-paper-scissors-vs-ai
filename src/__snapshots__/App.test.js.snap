// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> The Player picks paper, set the player move to paper 1`] = `
<button
  className="paper-btn"
  onClick={[Function]}
>
  <i
    className="fa fa-hand-paper-o large"
  />
</button>
`;

exports[`<App /> The Player picks rock, set the player move to rock 1`] = `
<button
  className="rock-btn yoo"
  onClick={[Function]}
>
  <i
    className="fa fa-hand-rock-o large"
  />
</button>
`;

exports[`<App /> The Player picks scissors, set the player move to scissors 1`] = `
<button
  className="scissors-btn"
  onClick={[Function]}
>
  <i
    className="fa fa-hand-scissors-o large"
  />
</button>
`;
